.github/workflows/build-docker-image.yml

```diff
@@ -0,0 +1,106 @@
+      # By default, images should be built for the latest commit of a release-branch.
+      # This should only be used to build test images,
+      # and never for production.
+      # Images will be tagged with <commit_sha>-<tag>, to enforce that the image is not used in production.
+      # Since the `commit_sha` will never match semver (vX.Y.Z).
+      enable_alpha_build:
+        description: 'Allow custom build for any branch, default: false'
+        required: true
+        type: boolean
+        default: false
+
+jobs:
+  # Generate image tag using:
+  # - version inside Cargo.toml,
+  # - the commit SHA,
+  # - and the tag provided by the user.
+  generate_image_tag:
+    runs-on: ubuntu-latest
+    # Checks that the branch is a release branch, by checking that the semver
+    # should match: vX.Y.Z
+    steps:
+      - uses: actions/checkout@v4
+      - name: 'Print inputs'
+        run: |
+          echo "TAG=${{ github.event.inputs.tag }}"
+          echo "enable_patch_build=${{ github.event.inputs.enable_patch_build }}"
+          echo "enable_alpha_build=${{ github.event.inputs.enable_alpha_build }}"
+      - name: 'Verify release branch'
+        id: verify_release_branch
+        if: github.event.inputs.enable_patch_build == 'false' && github.event.inputs.enable_alpha_build == 'false'
+        run: |
+          # Get current branch name
+          echo "GITHUB_REF=$GITHUB_REF"
+          BRANCH_NAME=$(basename $GITHUB_REF)
+          echo "BRANCH_NAME=$BRANCH_NAME"
+          echo "BRANCH_NAME=$BRANCH_NAME" >> $GITHUB_ENV
+
+          # Check for semver
+          if [[ "$BRANCH_NAME" =~ ^release-[0-9]+\.[0-9]+$ ]]; then
+            echo "Branch $BRANCH_NAME is a valid release branch"
+          else
+            echo "Error: Branch $BRANCH_NAME is not a release branch"
+            exit 1
+          fi
+      - name: 'Get base version'
+        id: get_base_version
+        if: github.event.inputs.enable_alpha_build != 'true'
+        run: |
+          VERSION=$(grep -m 1 '^version' Cargo.toml | cut -d '"' -f 2)
+          echo "VERSION=$VERSION"
+          if [[ $VERSION == *"alpha"* ]]; then
+            echo "Error: Branch $GITHUB_REF has an alpha version"
+            exit 1
+          fi
+          echo "VERSION=$VERSION" >> $GITHUB_ENV
+      - name: 'Generate image tag'
+        id: generate_image_tag
+        run: |
+          # Get the commit SHA
+          COMMIT_SHA=$(echo $GITHUB_SHA | cut -c1-7)
+          echo "COMMIT_SHA=$COMMIT_SHA"
+
+          # Get the tag from the input
+          TAG=${{ github.event.inputs.tag }}
+          echo "TAG=$TAG"
+
+          # Generate the image tag
+          if [[ ${{ github.event.inputs.enable_alpha_build }} == 'true' ]]; then
+            # If alpha build is enabled, use the commit SHA and tag
+            IMAGE_TAG="$COMMIT_SHA-$TAG"
+          else
+            # Use the version from Cargo.toml, commit SHA, and tag
+            IMAGE_TAG="v$VERSION-$TAG-$COMMIT_SHA"
```

Comment:
`COMMIT_SHA` goes first in alpha build while it becomes `TAG` in release build. Shall we normalize this?

Reply:
This is kind of a hack to prevent patching alpha build versions to production cluster in Rw Cloud. Since cloud will do semantic version checks, and alpha build versions will fail that if they use this workflow.

Reply:
Perhaps we can enhance the check on the cloud side rather than here

Reply:
Fixed, I removed both checks. I think it's straightforward to spot `alpha` in the version, and I included `branch` in the image tag as well, so we can verify the source branch elsewhere.

This pipeline will just focus on automating the image tag generation to keep it simple and stupid

Reply:
https://buildkite.com/risingwavelabs/docker/builds/17282
https://github.com/risingwavelabs/risingwave/actions/runs/14807210370/job/41577305338